### PR TITLE
Use separate URLClassPath targets depending on the JDK version #5335

### PR DESCRIPTION
**Pull request by galderz**
Thx @bobmcwhirter for pointers 😃

### CODE REVIEW

**Review by gastaldi — DISMISSED** [review]

File: extensions/vertx-core/runtime/src/main/java/io/quarkus/vertx/core/runtime/graal/JdkSubstitutions.java [hunk 1]
```diff
@@ -8,24 +8,42 @@
...
+@TargetClass(className = "sun.misc.URLClassPath$FileLoader", onlyWith = JDK8OrEarlier.class)
 final class Target_sun_misc_URLClassPath$FileLoader {
 
     @Alias
     public Target_sun_misc_URLClassPath$FileLoader(URL url) throws IOException {
     }
 }
 
-@TargetClass(className = "sun.misc.URLClassPath")
+@TargetClass(className = "jdk.internal.loader.URLClassPath$FileLoader")
```
Comment (gastaldi) [inline extensions/vertx-core/runtime/src/main/java/io/quarkus/vertx/core/runtime/graal/JdkSubstitutions.java]: This class doesn't exist in JDK8, is that missing a `onlyWith=JDK11OrLater.class`?

**Review by gastaldi — CHANGES_REQUESTED** [review]
Build Failed:

```
Error: substitution target for io.quarkus.vertx.core.runtime.graal.Target_jdk_internal_loader_URLClassPath$FileLoader is not loaded. Use field `onlyWith` in the `TargetClass` annotation to make substitution only active when needed.
com.oracle.svm.core.util.UserError$UserException: substitution target for io.quarkus.vertx.core.runtime.graal.Target_jdk_internal_loader_URLClassPath$FileLoader is not loaded. Use field `onlyWith` in the `TargetClass` annotation to make substitution only active when needed.
	at com.oracle.svm.core.util.UserError.abort(UserError.java:65)
	at com.oracle.svm.hosted.substitute.AnnotationSubstitutionProcessor.findTargetClass(AnnotationSubstitutionProcessor.java:823)
	at com.oracle.svm.hosted.substitute.AnnotationSubstitutionProcessor.handleClass(AnnotationSubstitutionProcessor.java:252)
	at com.oracle.svm.hosted.substitute.AnnotationSubstitutionProcessor.init(AnnotationSubstitutionProcessor.java:230)
	at com.oracle.svm.hosted.NativeImageGenerator.createDeclarativeSubstitutionProcessor(NativeImageGenerator.java:875)
	at com.oracle.svm.hosted.NativeImageGenerator.setupNativeImage(NativeImageGenerator.java:824)
	at com.oracle.svm.hosted.NativeImageGenerator.doRun(NativeImageGenerator.java:524)
	at com.oracle.svm.hosted.NativeImageGenerator.lambda$run$0(NativeImageGenerator.java:444)
	at java.util.concurrent.ForkJoinTask$AdaptedRunnableAction.exec(ForkJoinTask.java:1386)
	at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)
	at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056)
	at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692)
	at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:157)
Error: Image build request failed with exit status 1
```

See comment above

**Comment by bobmcwhirter** [comment]
Or use the `classNameProvider` annotation field https://github.com/oracle/graal/blob/master/substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/annotate/TargetClass.java#L103 if the substitution is identical between JDK9 and JDK11, and only the name/package of the class differs.

**Comment by gwenneg** [comment]
This PR fixes #5335, right?

**Comment by galderz** [comment]
@gwenneg Indeed it fixes it (or tries to), the commit contains that github issue id :)

@bobmcwhirter's suggestion seems better, I'll update the fix.

**Comment by galderz** [comment]
Pushed an update. Note that `URLClassPath` still has different classes because `getLookupCache` method only exists in the Java 8 (or earlier) version.

**Comment by galderz** [comment]
@gastaldi Was waiting for CI to finish first, is that Ok?

**Comment by galderz** [comment]
CI passed, can we integrate this?

**Comment by gastaldi** [comment]
@galderz sure, please squash the commits and I'll merge it. Thanks!

**Comment by galderz** [comment]
Ah yes, squashed

**Review by gastaldi — APPROVED** [review]

**Comment by gsmet** [comment]
I will not backport that one. There is a whole set of patches we need for GraalVM 19.3 support.

**Pull request merged**